File: backend/models/product.go
// Autores:
//   - Jennyfer Jasso
//   - Carlos Zamudio
//   - Min Che Kim

package models

// Estructura de la tabla de productos
type Product struct {
	ID uint `json:"id" gorm:"primarykey" validate:"omitempty,eq=0"`

	// ImageN
	Hash string `json:"hash"`
	Type string `json:"type"`

	// General
	Model       string  `json:"model"`
	Name        string  `json:"name"`
	Description string  `json:"description"`
	Price       float64 `json:"price"`
	Stock       int     `json:"stock"`

	// Especificaciones
	Size        string `json:"size"`
	Color       string `json:"color"`
	Maintenance string `json:"maintenance"`
	Material    string `json:"material"`
	Absorbency  string `json:"absorbency"`
	MatFeature  string `json:"material_feature"`

	// Relaciones
	CategoryID uint `json:"category_id"`
}

// Estructura de la tabla de los metadatos de los productos
type ProductMetadata struct {
	// General
	Model       string  `json:"model" validate:"required,min=1"`
	Name        string  `json:"name" validate:"required,min=1"`
	Description string  `json:"description" validate:"required"`
	Price       float64 `json:"price" validate:"required,min=0"`
	Stock       int     `json:"stock" validate:"min=0"`

	// Especificaciones
	Size        string `json:"size" validate:"required"`
	Color       string `json:"color" validate:"required"`
	Maintenance string `json:"maintenance" validate:"required"`
	Material    string `json:"material" validate:"required"`
	Absorbency  string `json:"absorbency" validate:"required"`
	MatFeature  string `json:"material_feature" validate:"required"`

	// Relaciones
	CategoryID uint `json:"category_id" validate:"required,min=0"`
}

type MostSellProducts struct {
	ID         uint   `json:"id"`
	Name       string `json:"name"`
	OrderCount uint   `json:"order_count"`
}

type CategoryProductCount struct {
	ID           uint   `json:"id"`
	CategoryName string `json:"category_name"`
	ProductCount uint   `json:"product_count"`
}

// Establece el ID del producto
func (p *Product) SetID(id uint) {
	p.ID = id
}

// Obtiene el hash del producto
func (p *Product) GetHash() *string {
	return &p.Hash
}

// Obtiene el tipo del producto
func (p *Product) GetType() *string {
	return &p.Type
}

// Establece los metadatos del producto
func (p *Product) SetMetadata(metadata *ProductMetadata) {
	if metadata == nil {
		return
	}

	p.Model = metadata.Model
	p.Name = metadata.Name
	p.Description = metadata.Description
	p.Price = metadata.Price
	p.Stock = metadata.Stock
	p.Size = metadata.Size
	p.Color = metadata.Color
	p.Maintenance = metadata.Maintenance
	p.Material = metadata.Material
	p.Absorbency = metadata.Absorbency
	p.MatFeature = metadata.MatFeature
	p.CategoryID = metadata.CategoryID
}

File: app/app/src/main/java/mx/cazv/todasbrillamos/model/models/Product.kt
package mx.cazv.todasbrillamos.model.models

/**
 * Clase de datos que representa un producto sin procesar.
 * @author Carlos Zamudio
 *
 * @property id El ID del producto.
 * @property hash El hash de la imagen del producto.
 * @property type El tipo de imagen del producto.
 * @property model El modelo del producto.
 * @property name El nombre del producto.
 * @property description La descripción del producto.
 * @property price El precio del producto.
 * @property stock La existencia del producto.
 * @property size El tamaño del producto.
 * @property color El color del producto.
 * @property maintenance El mantenimiento del producto.
 * @property material El material del producto.
 * @property absorbency La absorbencia del producto.
 * @property material_feature Las características del material del producto.
 * @property category_id El ID de la categoría del producto.
 */
data class ProductRaw(
    val id: Int,
    val hash: String,
    val type: String,
    val model: String,
    val name: String,
    val description: String,
    val price: Int,
    val stock: Int,
    val size: String,
    val color: String,
    val maintenance: String,
    val material: String,
    val absorbency: String,
    val material_feature: String,
    val category_id: Int
)

/**
 * Datos del producto
 * @author: Min Che Kim, Carlos Zamudio
 *
 * @property folder El nombre de la carpeta del producto.
 * @property product El producto.
 */
data class Product(
    val folder: String,
    val product: ProductRaw
)
